File: 198-house-robber/198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [0] * n
        for i in range(n):
            left = dp[i-2] if i -2 >= 0 else 0
            right = dp[i-1] if i -1 >= 0 else 0
            dp[i] = max(left+ nums[i], right)
        return dp[n-1]

File: 198-house-robber/198-house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        vector<int>dp(n);
        
        for (int i = 0; i < n; i++) {
            int left = (i > 2 ? dp[i - 3]: 0);
            int right = (i > 1? dp[i - 2]: 0);
            dp[i] = max(left, right) + nums[i];
        }
        return *max_element(dp.begin(), dp.end());
    }
};